[jit] Turn resolver into a class

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19239 [wip][jit] use fully qualified name for ScriptClasses
* #19238 [jit] Allow for segmented printing in PythonPrint
* #19237 [jit] add resolveType to Resolver
* **#19236 [jit] Turn resolver into a class**

Change `Resolver` from a closure to an interface that accepts multiple
implementations. No behavioral change in this PR, but this is necessary
to support different kinds of resolutions (I want to add "resolve to a
type").

gh-metadata: pytorch pytorch 19236 gh/suo/24/head

Differential Revision: [D14928292](https://our.internmc.facebook.com/intern/diff/D14928292)

suo committed:
[jit] Turn resolver into a class

suo committed:
Update on "[jit] Turn resolver into a class"

[jit] Turn resolver into a class

Change `Resolver` from a closure to an interface that accepts multiple
implementations. No behavioral change in this PR, but this is necessary
to support different kinds of resolutions (I want to add "resolve to a
type").

gh-metadata: pytorch pytorch 19236 gh/suo/24/head

suo committed:
Update on "[jit] Turn resolver into a class"

[jit] Turn resolver into a class

Change `Resolver` from a closure to an interface that accepts multiple
implementations. No behavioral change in this PR, but this is necessary
to support different kinds of resolutions (I want to add "resolve to a
type").

gh-metadata: pytorch pytorch 19236 gh/suo/24/head

suo committed:
Update on "[jit] Turn resolver into a class"

[jit] Turn resolver into a class

Change `Resolver` from a closure to an interface that accepts multiple
implementations. No behavioral change in this PR, but this is necessary
to support different kinds of resolutions (I want to add "resolve to a
type").

gh-metadata: pytorch pytorch 19236 gh/suo/24/head

suo commented:
@pytorchbot retest this please

zdevito commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -284,7 +284,8 @@ struct GraphFuser {
           script::CompilationUnit cu;
-          cu.define(source, script::nativeResolver, nullptr);
+          cu.define(
+              source, std::make_shared<script::NativeResolver>(), nullptr);
```
nit: I don't like std::make_shared, because it always creates bad error messages. Can we make a function `script::nativeResolver()` just return the shared_ptr?

zdevito commented on torch/csrc/jit/script/builtin_functions.cpp:
```diff
@@ -63,8 +63,9 @@ struct BuiltinFunctionRegistry {
  private:
   void loadSource(const std::string& source) {
     std::shared_ptr<CompilationUnit> cu = std::make_shared<CompilationUnit>();
+    auto resolver = std::make_shared<script::NativeResolver>();
```
... especially if we are going to do it more than once.

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -2772,32 +2772,46 @@ struct to_ir {
+  }
+
+ private:
+  const Resolver* otherResolver_;
```
This fails to forward type resolution calls to otherResolver_, which I think is wrong. I also wonder if this is a good pattern. What happens if we add another method to Resolver? How would someone know to update this proxy.

zdevito replied:
Maybe it is better to just move the function_table_ logic into to_ir rather than try to push it into the resolver.

zdevito replied:
Now I see that type resolution does not use this yet, so I assume it is not a bug yet, but it seems like a likely place where a bug could happen.

suo replied:
Yeah, wrapping the resolver will probably be fragile. I'll change

suo replied:
Actually, I changed my mind. Since `Resolver` is an abstract virtual, adding new methods to it will require implementations by inheritors, so it'll be hard to actually introduce a bug. Passing a function_table around feels unnecessary when we already have a way to do "look up this name in the outer environment".

zdevito commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -791,12 +798,12 @@ void initJitScriptBindings(PyObject* module) {
              ResolutionCallback rcb,
              bool has_self) {
-            c10::optional<Self> self;
+            auto resolver = std::make_shared<PythonResolver>(rcb);
```
Again, would be nice to have a function rather than std::make_shared.

zdevito reviewed (approved):
Looks good! Minor comments.

suo committed:
Update on "[jit] Turn resolver into a class"

[jit] Turn resolver into a class

Change `Resolver` from a closure to an interface that accepts multiple
implementations. No behavioral change in this PR, but this is necessary
to support different kinds of resolutions (I want to add "resolve to a
type").

gh-metadata: pytorch pytorch 19236 gh/suo/24/head
